# Module 8, Aggregate Functions

### Step 1: Write a query to find the number of unique actors in the actor table of the dvdrental database

In [1]:
SELECT COUNT(DISTINCT actor)
FROM actor;

(1 row(s) affected)

Total execution time: 00:00:00.012

count 200

### How many actors are present in the actors table?

### 200

### Step 2: Write a query to find the minimum rental amount in the payment table. Output the amount in a column labeled "Minimum Rental Amount".

In [2]:
SELECT MIN(amount) AS "Minimum Rental Amount"
FROM payment;

(1 row(s) affected)

Total execution time: 00:00:00.050

Minimum Rental Amount 0.00

### Step 3: Output the minimum, maximum, average, and most occuring rental amounts from the payment table. Create a unique title for each column. Round the average to the nearest two decimals.

In [5]:
SELECT MIN(amount) AS "Minimum Amount"
    , MAX(amount) AS "Maximum Amount"
    , round(AVG(amount),2) AS "Average Amount"
    , MODE() WITHIN GROUP (ORDER BY amount) AS "Most Common Amount"
FROM payment;

(1 row(s) affected)

Total execution time: 00:00:00.012

Minimum Amount Maximum Amount Average Amount Most Common Amount 0.00 11.99 4.20 4.99

### What was the difference between the average amount and the most common amount in the table?

0.79

### Step 4: Find the variance of the rental amounts across the payment table. Round the answer to the nearest two decimals.

In [8]:
SELECT round(VARIANCE(amount),2) AS VARIANCE
FROM payment;

(1 row(s) affected)

Total execution time: 00:00:00.008

variance 5.61

What is the variance in the rental amounts across the payment table?

  

### 5.61

### Step 5: Determine the correlation between the replacement cost of a dvd and its rental rate.

In [12]:
SELECT CORR(replacement_cost, rental_rate) AS Correlated_Rental_to_Replacement
FROM film;

(1 row(s) affected)

Total execution time: 00:00:00.009

correlated_rental_to_replacement -0.04461978289656093

### Is the correlation between the DVD's replacement cost and its rental rate positive or negative?

Negative

  

What does this say about the cost of replacement and the rental rate

  

### The rental rate is not driven by the cost of the replcement of the DVD. The correlation is only slightly negative, so it has some impact on the rental rate, but is not overly correlated to it.

### Step 6: Find the count of all films with a rental duration of at least 7 days

In [37]:
SELECT COUNT(film_id) AS film_count
FROM film
WHERE rental_duration >= 7;

(1 row(s) affected)

Total execution time: 00:00:00.010

film_count 191

### How many rentals were seven days or greater?

191

### Step 7: Find min and max rental rates and the mose common rental duration of each film title. Group the results by the most common number of rental days and the max rental rate. Limit the output to the top title by rental days and rate and include any additional tying rows that meet the criteria.

In [64]:
SELECT title
    , MIN(rental_rate):: FLOAT AS min_rental_rate
    , MAX(rental_rate)::FLOAT AS max_rental_rate
    , MODE() WITHIN GROUP (ORDER BY rental_duration) AS mode_of_rent_len
FROM film
GROUP BY title
ORDER BY mode_of_rent_len DESC, max_rental_rate DESC
FETCH FIRST 1 ROWS WITH TIES;


(62 row(s) affected)

Total execution time: 00:00:00.012

title min_rental_rate max_rental_rate mode_of_rent_len Games Bowfinger 4.99 4.99 7 Soldiers Evolution 4.99 4.99 7 Truman Crazy 4.99 4.99 7 Dangerous Uptown 4.99 4.99 7 Outlaw Hanky 4.99 4.99 7 Elizabeth Shane 4.99 4.99 7 Stepmom Dream 4.99 4.99 7 Bikini Borrowers 4.99 4.99 7 Right Cranes 4.99 4.99 7 Darn Forrester 4.99 4.99 7 Volume House 4.99 4.99 7 Darling Breaking 4.99 4.99 7 Slacker Liaisons 4.99 4.99 7 Vanilla Day 4.99 4.99 7 Chinatown Gladiator 4.99 4.99 7 Saturn Name 4.99 4.99 7 Trip Newton 4.99 4.99 7 Opposite Necklace 4.99 4.99 7 Records Zorro 4.99 4.99 7 Wash Heavenly 4.99 4.99 7 Desperate Trainspotting 4.99 4.99 7 Boiled Dares 4.99 4.99 7 Private Drop 4.99 4.99 7 Chasing Fight 4.99 4.99 7 Grinch Massage 4.99 4.99 7 Born Spinal 4.99 4.99 7 Chamber Italian 4.99 4.99 7 Lies Treatment 4.99 4.99 7 Wake Jaws 4.99 4.99 7 Bowfinger Gables 4.99 4.99 7 Trainspotting Strangers 4.99 4.99 7 Open African 4.99 4.99 7 Legally Secretary 4.99 4.99 7 Sleeping Suspects 4.99 4.99 7 Holes Brannigan 4.99 4.99 7 Holiday Games 4.99 4.99 7 Confessions Maguire 4.99 4.99 7 Brooklyn Desert 4.99 4.99 7 Wanda Chamber 4.99 4.99 7 Jet Neighbors 4.99 4.99 7 Bucket Brotherhood 4.99 4.99 7 Hollow Jeopardy 4.99 4.99 7 Coneheads Smoochy 4.99 4.99 7 Iron Moon 4.99 4.99 7 Impossible Prejudice 4.99 4.99 7 Igby Maker 4.99 4.99 7 Home Pity 4.99 4.99 7 Forrester Comancheros 4.99 4.99 7 Speed Suit 4.99 4.99 7 Flight Lies 4.99 4.99 7 Reds Pocus 4.99 4.99 7 Perdition Fargo 4.99 4.99 7 Coldblooded Darling 4.99 4.99 7 Musketeers Wait 4.99 4.99 7 Sensibility Rear 4.99 4.99 7 Fugitive Maguire 4.99 4.99 7 War Notting 4.99 4.99 7 Undefeated Dalmations 4.99 4.99 7 Peak Forever 4.99 4.99 7 Philadelphia Wife 4.99 4.99 7 Walls Artist 4.99 4.99 7 Control Anthem 4.99 4.99 7

### What is the top title with the highest rental rate and most common rental days over all rentals?

### Games Bowfinger

###   
How many total rentals meet the above criteria?

### 62

### Step 8: Write a query to find the percentage of missing values in the return\_date column of the film table. Ouput the result as an integer with the "%" character appended to the result

In [95]:
SELECT (SUM(CASE WHEN return_date IS NULL THEN 1 ELSE 0 END)::FLOAT/COUNT(*)*100)::INT || '%' AS percent_missing
FROM rental;

(1 row(s) affected)

Total execution time: 00:00:00.008

percent_missing 1%

### What percentage of the return\_dates are missing?

### 1%

### Step 9: Write a query to determine if the actor\_id column in the actor table enforces referential integrity

In [102]:
SELECT COUNT(DISTINCT actor_id) = COUNT(*) AS integrity_check
FROM actor;

(1 row(s) affected)

Total execution time: 00:00:00.014

integrity_check True

### Step 10: Write a query to the total rental duration for each film in the film table. Sort the results by title and rental duration from least to most alphabetically. Only return rows 22 through 30.

In [109]:
SELECT title
    , SUM(rental_duration)
FROM film
GROUP BY title, rental_duration
ORDER BY rental_duration
LIMIT 8
OFFSET 21;

(8 row(s) affected)

Total execution time: 00:00:00.005

title sum Kissing Dolls 3 Words Hunter 3 Range Moonwalker 3 Pajama Jawbreaker 3 Fight Jawbreaker 3 Redemption Comforts 3 Tuxedo Mile 3 Flintstones Happiness 3

### What is the title of the movie in the 22nd row?

### Kissing Dolls

### Step 11: Write a query to find the percentage of movies with an MPAA rating of 'G'. Name the column "Rating" and append "G rated" to the output.

In [115]:
SELECT COUNT(DISTINCT rating)::FLOAT/count(*)*100 || ' G rated' AS "Rating"
FROM film
GROUP BY rating
HAVING rating = 'G';

(1 row(s) affected)

Total execution time: 00:00:00.003

Rating 0.5617977528089888 G rated